DRILL-4653.json - Malformed JSON should not stop the entire query from progressing

ssriniva123 opened the pull request:
https://issues.apache.org/jira/browse/DRILL-4653

- The default is to stop processing as is today when JSON parser encounters an exception
- Setting store.json.reader.skip_malformed_records will ensure that query progresses after
skipping the bad records
- Added two unit tests
- Also did testing after deploying the new build: Both positive and negative tests were done.
- Negative test result:
org.apache.drill.common.exceptions.UserRemoteException: DATA_READ ERROR: Error parsing JSON - Unexpected character ('{' (code 123)): was expecting comma to separate OBJECT entries

amansinha100 commented:
A high level comment I have about this PR is that keeping the number of 'malformed' records in the json file may not be sufficient..I would expect the user would want to know the file name and the location within the file for such record.

ssriniva123 commented:
Do u anticipate this to be logged using log4j?

amansinha100 commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/json/JSONRecordReader.java:
```diff
@@ -64,6 +64,8 @@
   private final boolean enableAllTextMode;
   private final boolean readNumbersAsDouble;
   private final boolean unionEnabled;
+  private int parseErrorCount;
```
This should be long instead of int since parseErrorCount is cumulative..so in the worst case it could be as large as runningRecordCount.

parthchandra commented:
Does this conflict with Drill 3764 (see https://github.com/apache/drill/pull/497)?. I think there was a broader discussion around this as part of enabling the functionality of skipping/logging erroneous records.

amansinha100 commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/json/JSONRecordReader.java:
```diff
@@ -217,11 +228,11 @@ public int next() {
+   // }
     // this is never reached
-    return 0;
+    //return 0;
```
Uncomment this (best practice since function has a return type)

amansinha100 commented:
Yes, it does in fact have a conflict with DRILL-3764 which has changes to the JsonRecordReader, although this issue is still in progress.   I noticed that @adeneche  mentioned this in the JIRA. @ssriniva123 did you get a chance to look at DRILL-3764 ?

ssriniva123 commented:
Yes, I looked at DRILL-3764. The changes made to JSONProcessing do not conflict with my changes.
The only conflict I see is these two constants, I have defined my own.

 String ENABLE_SKIP_INVALID_RECORD_KEY = "exec.enable_skip_invalid_record";
 +  BooleanValidator ENABLE_SKIP_INVALID_RECORD = new BooleanValidator(ENABLE_SKIP_INVALID_RECORD_KEY, false);

ssriniva123 commented:
I have added support so that it prints the offending file/line nos.

Error parsing JSON in DRILL-4653.json : line nos :13
Error parsing JSON in DRILL-4653.json : line nos :14

ssriniva123 commented:
For some reason the build terminated- What needs to be done to fire it again?

ssriniva123 commented:
I updated the pull request with more changes requested by the reviewers.

On Wed, Jun 15, 2016 at 5:28 PM, Aman Sinha <notifications@github.com>
wrote:

> Yes, it does in fact have a conflict with DRILL-3764 which has changes to
> the JsonRecordReader, although this issue is still in progress. I noticed
> that @adeneche <https://github.com/adeneche> mentioned this in the JIRA.
> @ssriniva123 <https://github.com/ssriniva123> did you get a chance to
> look at DRILL-3764 ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/drill/pull/518#issuecomment-226357921>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/ABsaH_Xd1WBKpu396uYeUVluiQ4iN0_Rks5qMJihgaJpZM4IzZkt>
> .
>

amansinha100 commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/json/JSONRecordReader.java:
```diff
@@ -39,6 +40,7 @@
 import org.apache.drill.exec.vector.complex.fn.JsonReader;
 import org.apache.drill.exec.vector.complex.impl.VectorContainerWriter;
 import org.apache.hadoop.fs.Path;
+import org.apache.parquet.Log;
```
Not sure why the parquet.log is included in the json reader

amansinha100 commented on exec/java-exec/src/main/java/org/apache/drill/exec/ExecConstants.java:
```diff
@@ -135,6 +135,9 @@
   BooleanValidator JSON_EXTENDED_TYPES = new BooleanValidator("store.json.extended_types", false);
   BooleanValidator JSON_WRITER_UGLIFY = new BooleanValidator("store.json.writer.uglify", false);
   BooleanValidator JSON_WRITER_SKIPNULLFIELDS = new BooleanValidator("store.json.writer.skip_null_fields", true);
+  String JSON_READER_SKIP_MALFORMED_RECORDS_FLAG = "store.json.reader.skip_malformed_records";
```
Can you change this to 'skip_invalid_records' such that the name is somewhat consistent with the future similar option in DRILL-3764.  In the future the json option would likely be subsumed by the new global option.

amansinha100 commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/json/JSONRecordReader.java:
```diff
@@ -189,39 +194,37 @@ private long currentRecordNumberInFile() {
-        write = jsonReader.write(writer);
-
-        if(write == ReadState.WRITE_SUCCEED) {
+   // try
```
remove these commented lines

amansinha100 commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/json/JSONRecordReader.java:
```diff
@@ -189,39 +194,37 @@ private long currentRecordNumberInFile() {
-
-      jsonReader.ensureAtLeastOneField(writer);
-
+      catch(Exception ex)
```
minor style convention: can you put the catch() on the previous line to match the closing paren

amansinha100 commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/json/JSONRecordReader.java:
```diff
@@ -189,39 +194,37 @@ private long currentRecordNumberInFile() {
-
+      catch(Exception ex)
+      {
+           ++parseErrorCount;
```
the indentations seem to be off here as well as other places.. can you make sure the indentations match the rest of the code ?

amansinha100 commented on exec/java-exec/src/test/java/org/apache/drill/exec/store/json/TestJsonRecordReader.java:
```diff
@@ -116,6 +117,7 @@ public void testMixedNumberTypes() throws Exception {
           .jsonBaselineFile("jsoninput/mixed_number_types.json")
           .build().run();
     } catch (Exception ex) {
+      ex.printStackTrace();
```
not sure why this printStackTrace was added in a different test from the ones that you added...

parthchandra replied:
Not a good idea to print stack trace in unit tests. The output of our unit tests is already too verbose.
Use junit.fail with the message from the exception instead?

amansinha100 commented on exec/java-exec/src/test/java/org/apache/drill/exec/store/json/TestJsonRecordReader.java:
```diff
@@ -179,4 +181,28 @@ public void testNestedFilter() throws Exception {
+
+
+ @Test // See DRILL-4653
+  public void testSkippingInvalidJSONRecords() throws Exception {
```
For both these tests could you pls use the testBuilder() framework ?  This is the recommended way to write the unit tests .. you can see one of the other tests in this file.

ssriniva123 commented:
I have made changes are recommended by reviewers:

- Changed JSON_READER_SKIP_INVALID_RECORDS_FLAG constant
- Modified unit test to use builder framework
- Code indendation changes.

amansinha100 commented:
Looks much better.  Sorry for the nitpick but I still have a couple more related to the coding conventions. :)    Also, could you squash the commits into 1 and use the DRILL-XXXX: <commit message>  format for the commit ?   thanks !

amansinha100 commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/json/JSONRecordReader.java:
```diff
@@ -189,39 +191,33 @@ private long currentRecordNumberInFile() {
+    outside: while(recordCount < DEFAULT_ROWS_PER_BATCH){
+    try
+      {
+            writer.setPosition(recordCount);
```
seems this is still doing indent of 4.  We use 2 spaces (see https://drill.apache.org/docs/apache-drill-contribution-guidelines/   scroll down to Step 2).   Did it pass the mvn command line build without checkstyle violations ?

ssriniva123 replied:
Aman,
maven checkstyle:checkstyle did not report any errors before I did my last
check in. I have changed to reflect 2 spaces for indendation.

On Thu, Jun 16, 2016 at 2:22 PM, Aman Sinha <notifications@github.com>
wrote:

> In
> exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/json/JSONRecordReader.java
> <https://github.com/apache/drill/pull/518#discussion_r67426381>:
>
> > -      outside: while(recordCount < DEFAULT_ROWS_PER_BATCH) {
> > -        writer.setPosition(recordCount);
> > -        write = jsonReader.write(writer);
> > -
> > -        if(write == ReadState.WRITE_SUCCEED) {
> > -//          logger.debug("Wrote record.");
> > -          recordCount++;
> > -        }else{
> > -//          logger.debug("Exiting.");
> > -          break outside;
> > -        }
> > -
> > +    outside: while(recordCount < DEFAULT_ROWS_PER_BATCH){
> > +    try
> > +      {
> > +            writer.setPosition(recordCount);
>
> seems this is still doing indent of 4. We use 2 spaces (see
> https://drill.apache.org/docs/apache-drill-contribution-guidelines/
> scroll down to Step 2). Did it pass the mvn command line build without
> checkstyle violations ?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/drill/pull/518/files/56a16fe3f2bbd1554d65cef6faeaeb63ba41f9a2#r67426381>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABsaHw_13TKQS-3mIEDhxkCiglpCBA3Sks5qMb6IgaJpZM4IzZkt>
> .
>

amansinha100 commented on exec/java-exec/src/test/java/org/apache/drill/exec/store/json/TestJsonRecordReader.java:
```diff
@@ -179,4 +180,43 @@ public void testNestedFilter() throws Exception {
+public void testSkippingInvalidJSONRecords() throws Exception {
+    try
+    {
+    String set = "alter session set `" + ExecConstants.JSON_READER_SKIP_INVALID_RECORDS_FLAG+ "` = true";
```
these should be indented inside the try block with 2 spaces.   It is best to set the indent level in your IDE (I can help with Eclipse if you are using it;  if you are using IntelliJ I can find out from other developers using IntelliJ).

ssriniva123 commented:
I have also squashed my commits

amansinha100 commented:
hmm.. I still see 4 commits in the pull request.  Can you squash them into one ? (let me know if you need help with that).  Also, the commit message needs to be in the format:  "DRILL-4653: Malformed json...."

ssriniva123 commented:
What is the recommendation with this bug? Should we just remove the line nos for now and just name the file name. I was not able to work on this last couple of days, not sure if there is sufficient time until the next candidate release.

ssriniva123 commented:
I have made several modifications to get accurate line nos:

- The main change is to move the json parser to the end of the current record being processed since
previously multiple exceptions were thrown.

- BaseJsonProcessor contains new method protected JsonExceptionProcessingState processJSONException().
-This is called by both CountingJsonReader and JsonProcessor whenever parser encounters a jackson parsing exception.
-I have also added a new system setting store.json.reader.print_skipped_invalid_record_number so that we can suppress printing of line numbers.
- Added more unit test cases.
- System tested with various combinations on a local drill bit.

jinfengni commented:
@chunhui-shi , I saw you made comments days ago. Can you pls take a look at the new patch to see if it addressed your comment? thx.

chunhui-shi commented on exec/java-exec/src/main/java/org/apache/drill/exec/vector/complex/fn/JsonReader.java:
```diff
@@ -110,21 +118,29 @@ public void ensureAtLeastOneField(ComplexWriter writer) {
-        // that contains non-integer fields will error out in any case.  Whereas, with
-        // allTextMode true, we are sure that all fields are going to be treated as varchar,
-        // so it makes sense to produce all the fields, and in fact is necessary in order to
+        // when allTextMode is false, there is not much benefit to producing all
```
Seems the line width changed here(line 121-132). Please reorganize the text. If possible keep the original text unchanged. The same in line 140-143

chunhui-shi commented on exec/java-exec/src/test/java/org/apache/drill/exec/store/json/TestJsonRecordReader.java:
```diff
@@ -56,11 +57,19 @@ public void trySimpleQueryWithLimit() throws Exception {
     test("select a[0].b[0] from cp.`jsoninput/nestedArray.json`");
     test("select a[0].b[1] from cp.`jsoninput/nestedArray.json`");
-    test("select a[1].b[1] from cp.`jsoninput/nestedArray.json`");  // index out of the range. Should return empty list.
+    test("select a[1].b[1] from cp.`jsoninput/nestedArray.json`"); // index out
```
comment format

chunhui-shi commented on exec/java-exec/src/test/java/org/apache/drill/exec/store/json/TestJsonRecordReader.java:
```diff
@@ -159,64 +164,91 @@ public void drill_3353() throws Exception {
+  public void testCountingQueryNotSkippingInvalidJSONRecords() throws Exception {
+    try {
+      String query = "select count(*) from cp.`jsoninput/drill4653/file.json`";
+      testBuilder().unOrdered().sqlQuery(query).sqlBaselineQuery(query).build()
```
Should we compare with expected count here?

chunhui-shi commented:
Thanks for providing a way to skip bad json records and addressing my comments.
Besides the minor improvements I suggested above: comment format changes that may be due to line width in your env.,  and unit tests that should check the expected output as much as possible, could you please squash these multiple commits into one commit?

jaltekruse commented:
I don't think we should merge this without a mechanism to return a warning to the user to tell them at least that some data was ignored, and ideally some indication of how much data was discarded. While I do understand this is not the default behavior, I think there is still too high of a risk that an admin could set this at a global level and users would be unaware of some of their data being discarded.

I am willing to discuss the benefits of merging this before such a system exists, but until this issue has been thoroughly evaluated I am -1 on the change.

One improvement you could make to the current implementation is moving the option to the format plugin instead of the system/session list. This enables users to include setting the option in there query with the "table with options" syntax that was added last fall. We already have a JIRA open for moving the all_text_mode and read_numbers_as_double options to this location, because it doesn't really make sense to change query results based on session state. Unfortunately this change does not completely remove my initial concern, because not all users can modify or see the storage plugins in the case when web UI security is enabled. Non-admin users in these cases could be surprised by this behavior.

For examples of how this is done, you can look at the text plugin config, you would just need to add these options as properties to the json config which is currently mostly empty.
https://github.com/apache/drill/blob/master/exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/json/JSONFormatPlugin.java#L93

https://github.com/apache/drill/blob/master/exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/text/TextFormatPlugin.java#L135

Select with options: https://issues.apache.org/jira/browse/DRILL-4047
Jira for moving the existing options: https://issues.apache.org/jira/browse/DRILL-4206

jaltekruse commented:
After this was discussed at the Hangout a few weeks back, I had been thinking about it more.

My initial request was for warnings to be returned along with query results. Some initial work was posted last fall to add warnings to the RPC protocol, but unfortunately it was not brought to completion. These kinds of warnings can be received by JDBC and ODBC sources and many tools show them to users.

https://github.com/apache/drill/pull/263

In the hangout we discussed that this changeset is currently logging when part of a file is ignored. I don't believe that there is currently an expectation that users should have to grep through a log file to find out any additional information about the execution of a successful query, the log files are there for admins to debug system issues.

I still think it would be good to have a way to remove the need to look through the log file to see if this behavior was used in a query, but as there wasn't a lot of concern expressed by others when we discussed it, I'm changing my vote to a +0.

ssriniva123 commented:
Folks,
I have tests on my local drill setup with various combinations of invalid json format. The only caveat is that some records may be skipped if JSON is not properly delimitted by }. Can this be reviewed for next release ? I can work on the documentation.

paul-rogers commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/json/JsonProcessor.java:
```diff
@@ -30,6 +30,8 @@
 
   public static enum ReadState {
     END_OF_STREAM,
+    JSON_RECORD_PARSE_ERROR,
```
Would be helpful to add a comment to describe the meaning of these new states.

paul-rogers commented:
Thanks much for your contribution! Sorry it is taking a while to review the code.

I've read over the code changes a couple of times. The options part is fine. However, the parser part seems a bit too complex and may not handle all cases of interest. I worry that the code may only handle the error shown in the test input file: a missing close quote:

{ "foo" : "bar }
{ "foo" : "mumble" }

To do a more general recovery, we have to have some known good recovery point. Pure JSON is not suited to recovery. Fortunately, the JSON we read is not true JSON, rather it is a file of JSON objects. The file-of-json-object structure offers two possible recovery points.

If we require that each JSON record start on a new line, then we can use newline as our recovery point:

{ "foo" :: 10 }
{ "foo" : 20 }

In the above, the first record is badly formed. But, we know where the next record begins because of the newline separator.

It seems that the code may be using the above approach. But, since the only test case is a missing close quote, the parser will "naturally" read all tokens up to EOL. But, the implementation seems to not handle the double-colon example: the paser will fail on the second token. When resuming, the parser will see another colon and fail again. Maybe the parser would eventually fail enough to find the EOL. But, we can construct cases where things would fail:

{ "foo" :: 10, "x": { "bar", 20, "mumble" } }

If we don't discard tokens to the newline, the parser may decide that { "bar" ... starts a new valid record, with the result of causing a schema change when not needed.

The message here is that, on error, we must discard tokens to EOL, we can't just try to resume parsing at the point of failure.

If, however, we can't count on the newline, then we need some other syntactic trick. Perhaps we can look for "}/s*{" (that is, close bracket, optional white space, open bracket):

{ "foo" :: 
{ "foo" : 20 }
{ "foo" : 30 }

Here, we'd actually discard both the first and second records because we'd only find the recovery patttern between the second and third records. (The "} {" pattern can never occur inside a JSON object, it would have to be "}, {" instead.)

Using EOL as delimiter is easiest: we don't need lookahead. But, we require that a newline separate records. (Newlines within records can be handled, but we'll omit that for now...)

Using "}/s*{" as the recovery is a bit harder: we need to read ahead by one token, then push the token back on the input.

If we do the above, then the reader algorithm should look something like this:

    while ( more records to read ) {
        try {
            read the record
        } catch ( parse exception ) {
            read and discard tokens until we get to an EOL (or the "}/s*{")
            reset the value vector pointer back one step to discard any partially loaded values.
        }
    }

The above can be shown to be correct by working through a simple state machine and set of examples.

Alternatively, it might help to include here (or in comments in code) an explanation of the proposed recovery mechanism so it is easier for reviewers to verify correctness.

paul-rogers commented:
Upon reflection, it seems that newline is not an adequate marker to separate JSON records. Many of our samples have internal newlines. If a newline appears inside the JSON record, then we are subject to the same incorrect recovery as illustrated with the "a, x, bar, y" example in the earlier comment.

Further, if the JSON tokenizer is like most, it probably discards whitespace, not returning EOL as a token.

So, it seems that the best (or only) option is to scan for the "} {" pair. This requires two specific improvements:

* A "token discarder" that uses a state machine to look for the "} {" pairs, and
* An indirection around the get-token method so we can push the "{" token back onto the input.

These changes, along with the pseudo-code shown earlier may provide as good a solution as we can get. (Phrased that way because some errors will cause two records to be discarded, as explained earlier.) Combine that with the options and error reporting from the original pull request and we are probably pretty close.

ssriniva123 commented:
Paul,
Thanks for taking the time out in writing out a detailed email. Here are some of my thoughts.

- Drill uses the com.fasterxml.jackson.core.json.UTF8StreamJsonParser for parsing of JSON records. This parser does not rely on line delimiters for record separators but instead uses
the JSON structure as a natural way to signal End of record (EOR). There are methods internal
to the parser which check for line feeds but is not exposed to callers.

- The CountingJsonReader uses the parser.skipChildren() method to skip the rest of the children for this record, hence it is not possible to accurately count and match the no of braces to cleanly skip that bad record.

- One thought is to tap the inputsource of the parser on an exception condition, but is not
encouraged.

My thought process was exactly along the lines you have been thinking. On an exception scenario the code attempts to locate a closing bracket(}) followed by a opening bracket ({).
This is what is being done in the BaseJsonProcessor.processJSONException method. Please note that it works in all cases except when we do not have proper brackets to signify end of a JSON record. 

Hope this explanation helps clarify.

paul-rogers commented:
Poking around in the code a bit more, it looks like we rely on the Jackson JsonParser which delivers a stream of tokens. Here's the code in JsonReader:

    public ReadState write(ComplexWriter writer) throws IOException {
      JsonToken t = parser.nextToken();

As you pointed out, white space is consumed internally to the tokenizer. But we do get JSON tokens ({, }, identifier, number, etc.) Given this, you can scan ahead to find the close-open bracket pattern.

The trick seems to solve two problems. First, how do we handle tokens: The answer seems to be a few lines down:

    ReadState readState = writeToVector(writer, t);

Given a start token, we write to vector. writeToVector is a recursive-descent parser that consumes tokens and does the right thing. Here, it accepts only a top-level object, rejecting all other tokens. For an object, we call writeDataSwitch (funny name, that) which recursively writes fields, objects, lists and the rest.

Since this is a recursive-descent parser, JSON structure is represented by the call stack. To bail out of an invalid parse, we have to unwind the stack, which is what an exception does. So, that part may be good.

The next step is how to get the tokenizer (JsonParser) pointed at the right point so we can try to read the next object. That is where we need to consume tokens until we find the end-start bracket pair.

But, note that we have now consumed the start bracket, so we can't read it again when we gain call writeToVector. Checking the code, the JsonParser has no unget( ) method, unfortunately. To work around that, we need to "push" the token back onto the input. (Either actually doing so, or by having internal state that says that we've already read the open bracket.) 

We also have to ask if the JSON parser keeps track of parse state. Looking at the code, it seems that JsonParser is really just a tokenizer: it has not state about the JSON structure. (Would be good to run a small test to verify this observation.)

By the way, the JSON parser class has lots of good stuff already there. For example, the parser itself will keep track of line numbers and file locations. Perhaps we can use that when reporting error positions.

The last bit is that we've been building up a record as we parser JSON. If we fail part way thorugh, we've got a half-build record. Again, here I'm a bit hazy on what Drill can do. Can we "unwind" the current record? Can we mark the record as one to ignore (with a select vector?) Or, do we live with the half-build record? Throwing away the half-built record would be best, if we can do it.

All that said, how much of the above does your proposed code change handle? What other parts might still need to be added?

Thanks!

ssriniva123 commented:
Apologize for getting back on this thread late, got tied up with some issues@work.

Paul,
The json parser is not just a tokenizer, it keeps track of the JSON structure and understands various aspects of it like root, array/objectcontext and all parsing is done under that context.

- we cannot keep track of {} accurately - For eg: The counting json processor does a parser. skipChildren which tries to skip to the end of the JSON, but this can rollover to next line when
there is a malformed JSON in the bottom most json sub object - see example below (missing " in last json structure). This is similar behavior with the JsonReader.

{"balance": 1000.0,"num": 100,"is_vip": true,"name": "foo3","curr":{"denom":"pound","test":{"value  :false}}}

- One possible solution is to rewind the input source to reset the stream (which is not recommended and there is no guarentee that all streams support mark/reset semantics.

Given where we are, I think the solution proposed works perfect for almost all malformed JSON's.

paul-rogers commented:
Looks like you are right; the JsonParser is more than a simple tokenizer.

We're not the first to try this: http://stackoverflow.com/questions/37511496/recover-from-malformed-json-with-jackson (no answer)

I tried an experiment and found that you are on the right track: the way you are using the JsonParser can be extended to ignore input until the start of the next object. A quick demonstration:

    private static void recover(JsonParser parser) throws IOException {
      for ( ; ; ) {
        JsonToken token;
        try {
          token = parser.nextToken();
        } catch( JsonParseException e ) { continue; }
        if ( token == null ) return;
        if ( token != JsonToken.END_OBJECT ) { continue; }
        token = parser.nextToken();
        if ( token == null ) return;
        if ( token == JsonToken.START_OBJECT ) { return; }
      }
    }

Basically, we keep reading tokens, and ignoring errors, until we successfully find the } { pair.

As we discussed before, to use the above in Drill, we have to discard the partly-built record, and start reading the next record assiming the parser is positioned **after** the START_OBJECT ("{") token, which we've already consumed. That should be simple.

Still, to do proper recovery, we have to discard the partly-built JSON record. I've not looked into how to do that. If we don't do that, we return the bogus partly-built record. Worse, if we recover by trying to build a new record, we create more partly-built records, but with a different schema, possibly triggering a schema change event when not really necessary.

Any ideas for how to solve that problem?

ssriniva123 commented:
There is not much to do except change the JSON parser to support this functionality.

- Indicate to the parser that a current record terminates when it encounters a \n (Of course
this then assumes that druid also aligns to record separators using a new line).
This is a change to make to the jackson parser.

- Right now the current code works for all the standard cases except one case where the inner 
most sub-structure within a JSON is malformed.

Given that this is a great recipe for approximation algorithms , I am requesting this change to be pulled in.

If need be we can work on change to the jackson parser using a different request.

paul-rogers commented:
As it turns out, the sample code shown was actually tested with a stock Jackson JSON parser: it does work. No parser changes are needed.

The issue is not whether we can make the parser do what is needed: the code posted in the comment above demonstrated a solution.

The issue is how we incorporate that code into the JSON parser to clean up partial records and prevent schema changes. When I have time, I'll investigate that question in greater depth.

IMHO, without a proper fix, we should simply state that Drill does not support malformed JSON. If an input file might be incorrect, run it though a clean-up step before allowing Drill to scan it. Otherwise, we are opening the door to many hard-to-resolve bugs when people ask Drill to scan corrupt JSON: the result, without a proper fix, would be undefined -- which is worse than the current behavior that simply fails the scan with an error.

Let's follow up again after I (or someone) has had a chance to figure out if we can undo a partially built record. If we can do that, then we've got a path to a clean solution: recover the parser (as shown earlier) and discard the in-flight record (as we need to research.)

paul-rogers commented:
The open question was how we can discard a partly-built record during recovery. As far as I can tell (veterans, please correct me), the JSONRecordReader keeps track of the record count. So, all we have to do is not increment the count when we want to discard a record. Look in

    JSONRecordReader.next( )
      ...
      outside: while(recordCount < DEFAULT_ROWS_PER_BATCH) {
        writer.setPosition(recordCount); // Sets the position for the next read.
        write = jsonReader.write(writer); // Write the record. We can catch errors
                                          // and recover here??
         ...
          recordCount++; // Don't do this on a bad record
      ...
      writer.setValueCount(recordCount); // The record reader controls the record count.

This seems to show the elements of a solution:

1. Try to read the record.
2. If a failure occurs, catch it here and clean up, as in the previous post.
3. Don't increment the record count. We reuse the current one on the next record read.

Now the only open question is how we clean up the in-flight record in case some columns are not present in the next record. Anyone know how to set a vector position to null (for optional) default value (for required) or zero-length (for repeated)?

ssriniva123 commented:
Paul,
The code you have listed is semantically equivalent to that of what I already I have submitted for pull and will not solve handling of all malformed json records. Also the code for reporting the 
error records is working correctly as long as is it is reported by the Parser correctly.

As I explained earlier the JSON parser is not just a simple tokenizer, it keeps track of internal state,
hence the issue. SERDE's in hive etc work because they  are record oriented with clean record demarkations using a new line.

One solution is to submit a patch to jackson parser to expose a method to skip to new line in the
event of a parsing exception. This can be parametrized so that behavior can customized.

paul-rogers commented:
Ran some tests. The results look good. In particular, files with nested structures produced the correct results. Since it was the nested structure case that had me a bit worried, looks like the code is good to go.

+1 (non-binding)

zfong commented:
@paul-rogers - since you had concerns about particular test cases, and it looks like you've confirmed that those are non-issues, would it make sense for you to share those with @ssriniva123 and he can then include them as unit tests with this pull request?

paul-rogers commented:
Base data set:

    { "p": 1, "a": { "x": 10, "y": 20, "z": 30 }, "b": 50, "c": 60 }
    { "p": 2, "a": { "x": 11, "y": 21, "z": 31 }, "b": 51, "c": 61 }
    { "p": 3, "a": { "x": 12, "y": 22, "z": 32 }, "b": 52, "c": 62 }
    { "p": 4, "a": { "x": 13, "y": 23, "z": 33 }, "b": 53, "c": 63 }
    { "p": 5, "a": { "x": 14, "y": 24, "z": 34 }, "b": 54, "c": 64 }
    { "p": 6, "a": { "x": 15, "y": 25, "z": 35 }, "b": 55, "c": 65 }

Create various errors:

    { "p": 2x, "a": { "x": 11, "y": 21, "z": 31 }, "b": 51, "c": 61 }

and

    { "p": 2, "a": { "x": 11x, "y": 21, "z": 31 }, "b": 51, "c": 61 }

And so on.

In running the tests, I consistently saw that only the second (bad) row was omitted. Other rows properly appeared, and no partial row appeared.

ssriniva123 commented:
I have a similar data set checked in. Hope that is good enough.


On Fri, Oct 7, 2016 at 10:00 AM, Paul Rogers <notifications@github.com>
wrote:

> Base data set:
>
> { "p": 1, "a": { "x": 10, "y": 20, "z": 30 }, "b": 50, "c": 60 }
> { "p": 2, "a": { "x": 11, "y": 21, "z": 31 }, "b": 51, "c": 61 }
> { "p": 3, "a": { "x": 12, "y": 22, "z": 32 }, "b": 52, "c": 62 }
> { "p": 4, "a": { "x": 13, "y": 23, "z": 33 }, "b": 53, "c": 63 }
> { "p": 5, "a": { "x": 14, "y": 24, "z": 34 }, "b": 54, "c": 64 }
> { "p": 6, "a": { "x": 15, "y": 25, "z": 35 }, "b": 55, "c": 65 }
>
> Create various errors:
>
> { "p": 2x, "a": { "x": 11, "y": 21, "z": 31 }, "b": 51, "c": 61 }
>
> and
>
> { "p": 2, "a": { "x": 11x, "y": 21, "z": 31 }, "b": 51, "c": 61 }
>
> And so on.
>
> In running the tests, I consistently saw that only the second (bad) row
> was omitted. Other rows properly appeared, and no partial row appeared.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/drill/pull/518#issuecomment-252306011>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABsaH1IiZY7CSrdvx17MdDLrIyMFizqeks5qxnrDgaJpZM4IzZkt>
> .
>

kfaraaz commented:
The below JSON is invalid, due to presence of duplicate key 'key'. Today Drill returns a DATA_READ error, does your proposed fix handle this case too ?
[root@centos-01 ~]# cat f1.json
{"key":"string", "key":123, "key":[1,2,3], "key":true, "key":false, "key":null, "key":{"key2":"b"}}

Error returned by Drill 1.9.0

0: jdbc:drill:schema=dfs.tmp> select * from `f1.json`;
Error: DATA_READ ERROR: Error parsing JSON - You tried to write a BigInt type when you are using a ValueWriter of type NullableVarCharWriterImpl.

File  /tmp/f1.json
Record  1
Fragment 0:0

[Error Id: 06411bc5-2d59-4681-a84f-3f49086e18c0 on centos-01.qa.lab:31010] (state=,code=0)

parthchandra commented:
+1. Looks like there has been enough review and there is good enough reason to merge this in

paul-rogers commented:
Further testing revealed limitations in the underlying Jackson parser. That parser will not recover from other errors such as:
```
{ a: }
```
See [DRILL-5953](https://issues.apache.org/jira/browse/DRILL-5953) for details.
